File: S2-Regression/Pt2-MultipleLinearRegression/multiple_linear_regression.py
## Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd


## Importing the dataset
dataset = pd.read_csv('50_Startups.csv')
x = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values


## Encoding categorical data
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(
    transformers=[('encoder', OneHotEncoder(), [3])], remainder='passthrough')
x = np.array(ct.fit_transform(x))


## Avoiding the dummy variable trap
x = x[:, 1:]


## Splitting the dataset into training set and test set
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=.2, random_state=0)


## Trining Multiple Linear Regression Model on the training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(x_train, y_train)


## Predicting the test set results
y_pred = regressor.predict(x_test)
np.set_printoptions(precision=2)
print(np.concatenate(
    (y_pred.reshape(len(y_pred), 1), y_test.reshape(len(y_test), 1)), 1))


## Building the optimal model using Backward Elimination
import statsmodels.api as sm
x = np.append(arr=np.ones((50, 1)).astype(int), values=x, axis=1)
x_opt = x[:, [0, 1, 2, 3, 4, 5]].astype(np.float64)
x_opt = x[:, [0, 2, 3, 4, 5]].astype(np.float64)
x_opt = x[:, [0, 3]].astype(np.float64)
regressor_OLS = sm.OLS(endog=y, exog=x_opt).fit()
print(regressor_OLS.summary())

File: S2-Regression/Pt2-MultipleLinearRegression/multiple_linear_regression.R
## Importing the dataset
dataset = read.csv('50_Startups.csv')


## Encoding categorical data
dataset$State = factor(
  dataset$State, levels = c('New York', 'California', 'Florida'), labels = c(1, 2, 3)
)


## Splitting the dataset into training set and test set
# install.packages('caTools')
library('caTools')
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 3/4)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)


## Fitting Multiple Linear Regression to the training set
regressor = lm(formula = Profit ~ ., data = training_set)
summary(regressor)


## Predicting the test set results
y_pred = predict(regressor, newdata = test_set)
y_pred


## Building the optimal model using Backward Elimination
regressor = lm(
  formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
  data = dataset
)
regressor = lm(
  formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
  data = dataset
)
regressor = lm(
  formula = Profit ~ R.D.Spend + Marketing.Spend,
  data = dataset
)
regressor = lm(
  formula = Profit ~ R.D.Spend,
  data = dataset
)
summary(regressor)